subsys/net/ip/dhcpv4.c

```diff
@@ -880,14 +886,48 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
 
 			break;
 		}
-		default:
-			NET_DBG("option unknown: %d", type);
+		default: {
+			struct net_dhcpv4_option_callback *cb, *tmp;
+			struct net_pkt_cursor backup;
+			bool unhandled = true;
 
-			if (net_pkt_skip(pkt, length)) {
-				NET_DBG("option unknown, skip err");
-				return false;
+			net_pkt_cursor_backup(pkt, &backup);
+
+			SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&option_callbacks,
+							  cb, tmp, node) {
+				if (cb->option == type) {
+					__ASSERT(cb->handler, "No callback handler!");
+
+					cb->truncated = length > cb->max_length;
+
+					if (net_pkt_read(pkt, cb->data,
+							 MIN(cb->max_length, length))) {
+						NET_DBG("option callback, read err");
+						return false;
+					}
+
+					cb->handler(type, cb, MIN(cb->max_length, length));
```

Comment:
Another concern I have - currenly the option cb will be called for any message type. This could cause abiguity, in case there are more DHCP servers in the network (as theoretically we can receive mutliple OFFER messages). Perhaps the callback shoudl only be called on ACK messages?

Reply:
Wouldn't that concern apply to the way e.g. subnet mask and router are currently assigned? From what I can tell, they also don't have any extra checks in place.
Either way, it would be possible to pass `msg_type` to the callback.

Reply:
Please let me know if you believe the updated code sufficiently addresses this

Reply:
That is actually a very good point and something that may have been overlooked - probably because the ACK message would typically be the last one, but temporarily we could have got incorrect configuration stored in the system. 

As for the callback, I think that passing message type to the cb is also fine.